File: duna-core/src/main/java/io/duna/core/io/BufferOutputStream.java
/*
 * Copyright (c) 2017 Duna Open Source Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.io;

import io.vertx.core.buffer.Buffer;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.OutputStream;

public class BufferOutputStream extends OutputStream {

    private Buffer buffer;

    public BufferOutputStream() {
        this.buffer = Buffer.buffer();
    }

    public BufferOutputStream(Buffer buffer) {
        this.buffer = buffer;
    }

    public Buffer getBuffer() {
        return buffer;
    }

    @Override
    public void write(int bt) throws IOException {
        buffer.appendByte((byte) bt);
    }

    @Override
    public void write(@NotNull byte[] bytes) throws IOException {
        buffer.appendBytes(bytes);
    }

    @Override
    public void write(@NotNull byte[] bytes, int off, int len) throws IOException {
        buffer.appendBytes(bytes, off, len);
    }
}

File: duna-core/src/main/kotlin/io/duna/core/io/BufferOutputStream.kt
/*
 * Copyright (c) 2016 Duna Project
 * Ministério do Planejamento, Desenvolvimento de Gestão
 * República Federativa do Brasil
 *
 * This file is part of the Duna Project.
 */
package io.duna.core.io

import io.vertx.core.buffer.Buffer
import java.io.OutputStream

/**
 * Wraps a [io.vertx.core.buffer.Buffer] with an [java.io.OutputStream].
 *
 * @author Carlos Eduardo Melo [ceduardo.melo@gmail.com]
 * @since 0.1
 *
 * @see io.vertx.core.buffer.Buffer
 * @see java.io.OutputStream
 */
class BufferOutputStream(val buffer: Buffer) : OutputStream() {

  constructor(initialBufferSize: Int): this(Buffer.buffer(initialBufferSize))
  constructor(bytes: ByteArray): this(Buffer.buffer(bytes))

  override fun write(b: Int) {
    buffer.appendByte(b.toByte())
  }

  override fun write(b: ByteArray?) {
    buffer.appendBytes(b)
  }

  override fun write(b: ByteArray?, off: Int, len: Int) {
    buffer.appendBytes(b, off, len)
  }
}
